### DISCUSSION TITLE
On device actionable intents

### DISCUSSION THREAD

**Discussion by josephabbey** [home-assistant/architecture#1111] [General]
As it stands, the voice assistant flow is as follows:

```
Command --> Intent --> Action --> Response
```

Or for AI stuff:

```
Command --> (Processing...) Intents + Actions --> Response
```

I am building an Android App / Integration for HA that provides an alternative way to access assist, notably it uses the Android STT and TTS APIs and Material You theming.

It displays the contents of `targets`, `success`, and `failed` with the textual response.

I would like it to be able to perform actions like toggle the torch and navigate to an entity.

For example (Torch):

```
Turn on the torch
Turn off the torch
```

(Navigation):

```
Navigate me to Joseph
Navigate me home

Navigate me to 10 Downing Street
```

Both of these intents require the device itself to take action, not the HA Server itself.

## 3 possibilities

1. Send a message from the server to the app using something like Firebase Cloud Messaging or platform equivalent.
2. Return the required action as part of the response JSON.
3. Create a light entity on the HA server (only works for the torch).

My preference is option 2 (They are listed in the order that I thought of them).

The issue with option 2 is that it requires changes to conversation entities. Especially for AI entities, that can do multiple intents per command, they need to return all of the actions needed.

## Timers

This also raises questions about timers, the current way timers are implemented, from what I can tell, is option 1. This only works for devices with a solid, fast, bi-directional communication system with the server.

If we do chose option 2, will we change the timers implementation?

**Comment by allenporter** [comment]
Naively I would think from a core perspective, having core perform an action on a phone would be similar to having core take action on any other device. For example, an intent or service call targeting an entity or device.  

I read reading the notification commands in the developer docs and see there is prior art for example of sending a notification to run an Android Intent: https://companion.home-assistant.io/docs/notifications/notification-commands#activity and there are examples of launching other apps, so in that sense there is already an API for core to issue commands on the device through a service call.

**Reply by josephabbey** [reply]
Yes, in fact, none of what I am trying to accomplish here is not currently possible through a slightly clunky system. Your first paragraph details option 3 and your second details option 1. My preference of option 2 is purely to reduce total "clunk".

Rather than send a response and a separate command, they could be together. In my experience, my phone receives notification commands almost instantaneously, but my family's phones take up to 5 mins to receive them. The issue there is if they ask "turn on the torch", they want it on now, not at some point in the next 5 mins.

My original thought was to do exactly what you said and in fact I did implement it with an automation... and it did work. But I don't think it is at all preferable.

*I have also updated with a new option and caveats that I have just thought of.*

**Reply by allenporter** [reply]
Maybe worth spelling out a technical proposal of what the api would look like and how the complexity introduced would be managed. (E.g. is this introducing a new concept or reusing an existing one, how does core know what to do,, how are these actions registered next to the existing intents). 

Seems like there may be a big product component to this, and I am thinking this not a broad home assistant use case.

Edit: Corrected to say "not a broad home assistant use case"

**Reply by allenporter** [reply]
Oops -- in my response before I meant to say "not a broad home assistant use case" as in i'm not sure that there are a bunch of key product use cases for this.

I agree with you on major communication delays being a horrible user experience. That seems like a problem worth focusing on since it goes beyond voice actions if I understand right. Perhaps solving that and getting the best practices for reliable commands/notifications would be worthwhile (e.g. if its local push, changing settings to get more background time, or critical updates, etc).  Or rather, it seems worth solving/understanding anyway independent of this proposal as having flaky commands isn't good regardless.

**Reply by josephabbey** [reply]
I think that there is scope for quite a few commands that require action on the client where the server cannot send commands. Such as:
- View a map
- View a calendar
- View a list
- View an entity
- View a camera
- Navigate to an entity
- Turn on/off the torch
- Launch tag scanner
- Open an app/webpage (esp. ingress)
- Open a search
- Display a weather forecast
- Timers

Edit: I realise for the web frontend, this can be accomplished using the web socket API but for devices using just REST, it is not an option. Also changing settings to improve the notification speed that also reduce battery life or consume resources is not a preferable solution.

**Reply by josephabbey** [reply]
I welcome any other solutions, but for my problem, **so far**, the only working idea is option 2.

**Comment by josephabbey** [comment]
## Option 4:

Process the command on device first, to match on device requests.

This is not preferable as it is hard to handle languages and it does not work with conversation entities, esp. AI.

**Comment by josephabbey** [comment]
# Option 2:

> Return the required action as part of the response JSON.

Here is an example of a JSON response from a conversation request:

> ```json
> {
>   "response": {
>     "speech": {
>       "plain": {
>         "speech": "Turned off the lights",
>         "extra_data": null
>       }
>     },
>     "card": {},
>     "language": "en-GB",
>     "response_type": "action_done",
>     "data": {
>       "targets": [],
>       "success": [
>         {
>           "name": "Joseph's Bedroom",
>           "type": "area",
>           "id": "bedroom"
>         },
>         {
>           "name": "Joseph's Bedside Light",
>           "type": "entity",
>           "id": "light.joseph_s_bedside_light_switch"
>         },
>         {
>           "name": "Joseph's Bedroom Ambient Light",
>           "type": "entity",
>           "id": "light.joseph_s_bedroom_ambient_light_switch"
>         }
>       ],
>       "failed": []
>     }
>   },
>   "conversation_id": null
> }
> ```

It would make sense to change the `response.response_type` to `"action_needed"`, but I am weary of adding values to enums that span accross codebases and implementations.

We could also add an attribute to `response.data`:

```json
"data": {
  "actions": [
    {
      ...
    }
  ],
  ...
}
```

I don't know exactly what the contents of each action object will be, it will need some standardised (core) actions (like timers) and some custom actions.

Each device will need to be able to define which actions it can perform for intent recognition and to produce more accurate response text.

Another issue stems from the fact that conversation responses are sometimes produced/aggregated by conversation entities, so they will need to be changed to ensure that they pass on the actions.

On the intent handler side, the API would look like this:

```python
response: IntentResponse

response.async_set_actions(
    [
        IntentResponseAction(
            ...
        )
    ]
)
```

And the response would be returned as normal.

As it stands, the default conversation agent would not need to change.

The LLM API would have to change as well as the Gemini and ChatGPT conversation agents. Also there would need to be a way for custom intent handlers to tell the LLM API to ignore the intent if it is not supported.

**Reply by josephabbey** [reply]
There also needs to be a way to handle errors on the client side, if the action failed. This is a client side problem, though. It would be nice to have properly translated error messages, so maybe some error messages get passed along in the action object?
